### PR TITLE
Enable Ruff formatting

### PR DESCRIPTION
**Pull request by VeckoTheGecko**
Enable Ruff formatting for all Python files in codebase. Line length 88 for `docs/*`, 120 for the rest.

This PR is intended to isolate these changes (since they're non-breaking, but create a large diff), as well as to act as a place to flag any changes that are unreadable and would benefit from refactoring before being merged.

If a file is particularly bad, we can add them to an exclude list.

To disable the formatter for certain parts of the code, [you can add comments to suppress the formatter](https://docs.astral.sh/ruff/formatter/#format-suppression).

See related discussion in #1653. Contributes to #1620

### CODE REVIEW

File: parcels/fieldfilebuffer.py [hunk 1]
```diff
@@ -368,18 +415,29 @@ def _get_available_dims_indices_by_request(self):
...
+                    name not in self.chunksize
+                    or (
+                        type(self.chunksize[name]) is tuple
+                        and len(self.chunksize[name]) == 2
+                        and self.chunksize[name][1] <= 1
+                    )
+                )
+            ) or (
+                (type(self.chunksize) is tuple) and name in self.dimensions and (self.chunksize[i - tpl_offset] <= 1)
+            ):
```
Comment (VeckoTheGecko) [inline parcels/fieldfilebuffer.py:436]: I think this should probably be refactored or something

File: parcels/fieldfilebuffer.py [hunk 2]
```diff
@@ -433,10 +491,14 @@ def _is_dimension_chunked(self, dimension_name):
...
-        if self.dimensions is None or self.dataset is None or self.chunksize in [None, False, 'auto']:
+        if self.dimensions is None or self.dataset is None or self.chunksize in [None, False, "auto"]:
             return False
         dim_chunked = False
-        dim_chunked = True if (not dim_chunked and type(self.chunksize) is dict and dimension_name in self.chunksize.keys()) else False
+        dim_chunked = (
+            True
+            if (not dim_chunked and type(self.chunksize) is dict and dimension_name in self.chunksize.keys())
+            else False
+        )
```
Comment (VeckoTheGecko) [inline parcels/fieldfilebuffer.py:501]: Refactor large ternary

File: parcels/kernel.py [hunk 3]
```diff
@@ -377,20 +422,26 @@ def get_kernel_compile_files(self):
...
+                self._cache_key
+            )  # only required here because loading is done by Kernel class instead of Compiler class
             dyn_dir = get_cache_dir()
             basename = "%s_0" % cache_name
         lib_path = "lib" + basename
         src_file_or_files = None
         if type(basename) in (list, dict, tuple, ndarray):
-            src_file_or_files = ["", ] * len(basename)
+            src_file_or_files = [
+                "",
```
Comment (VeckoTheGecko) [inline parcels/kernel.py:443]: remove trailing comma

File: parcels/kernel.py [hunk 4]
```diff
@@ -484,7 +547,13 @@ def from_list(cls, fieldset, ptype, pyfunc_list, *args, **kwargs):
     def cleanup_remove_files(lib_file, all_files_array, delete_cfiles):
         if lib_file is not None:
             if os.path.isfile(lib_file):  # and delete_cfiles
-                [os.remove(s) for s in [lib_file, ] if os.path is not None and os.path.exists(s)]
+                [
+                    os.remove(s)
+                    for s in [
+                        lib_file,
+                    ]
```
Comment (VeckoTheGecko) [inline parcels/kernel.py:554]: remove trailing comma

File: parcels/fieldset.py
Comment (VeckoTheGecko) [inline parcels/fieldset.py:1]: Refactor to remove duplication in parameters (separate PR)

File: parcels/grid.py [hunk 5]
```diff
@@ -146,79 +176,113 @@ def lon_particle_to_target(self, lon):
...
-                    self.update_status = 'first_updated'
-            if signdt >= 0 and (self.ti < len(self.time_full)-2 or not f.allow_time_extrapolation):
-                nextTime_loc = self.time[1] + periods*(self.time_full[-1]-self.time_full[0])
+                    self.update_status = "first_updated"
+            if signdt >= 0 and (self.ti < len(self.time_full) - 2 or not f.allow_time_extrapolation):
+                nextTime_loc = self.time[1] + periods * (self.time_full[-1] - self.time_full[0])
             elif signdt < 0 and (self.ti > 0 or not f.allow_time_extrapolation):
-                nextTime_loc = self.time[0] + periods*(self.time_full[-1]-self.time_full[0])
+                nextTime_loc = self.time[0] + periods * (self.time_full[-1] - self.time_full[0])
         return nextTime_loc
```
Comment (VeckoTheGecko) [inline parcels/grid.py:286]: Can generally benefit from a refactor, but too large a task to do now

**Review by VeckoTheGecko — COMMENTED** [review]
Just using this PR draft to look through changes and note blocks where code formatting should be disabled (e.g., visualising 2D arrays, or long equations) or code should be refactored.

Will force push and rework the commits in this PR with the changes after I've some some refactoring in `fieldset.py` in another PR

File: parcels/application_kernels/TEOSseawaterdensity.py [hunk 6]
```diff
@@ -87,6 +88,34 @@ def PolyTEOS10_bsq(particle, fieldset, time):
...
+                + R020
+            )
+            * tt
+            + ((((R510 * ss + R410) * ss + R310) * ss + R210) * ss + R110) * ss
+            + R010
+        )
+        * tt
+        + (((((R600 * ss + R500) * ss + R400) * ss + R300) * ss + R200) * ss + R100) * ss
+        + R000
+    )
```
[the remark this answers is not in the document]
Comment (erikvansebille) [reply parcels/application_kernels/TEOSseawaterdensity.py:120]: Yes agree!

File: parcels/grid.py (hunk 5 above)
Comment (erikvansebille) [reply parcels/grid.py:286]: Note that grid.py may be completely rewritten when we start work on unstructured grids anyways

File: parcels/fieldfilebuffer.py (hunk 1 above)
Comment (VeckoTheGecko) [reply parcels/fieldfilebuffer.py:436]: Will leave for a future refactoring

File: parcels/fieldfilebuffer.py (hunk 2 above)
Comment (VeckoTheGecko) [reply parcels/fieldfilebuffer.py:501]: Will leave for a future refactoring

File: parcels/particledata.py [hunk 7]
```diff
@@ -64,19 +65,22 @@ def __init__(self, pclass, lon, lat, depth, time, lonlatdepth_dtype, pid_orig, n
 
         self._sorted = np.all(np.diff(pid) >= 0)
 
-        assert depth is not None, "particle's initial depth is None - incompatible with the ParticleData class. Invalid state."
-        assert lon.size == lat.size and lon.size == depth.size, ('lon, lat, depth don''t all have the same lenghts.')
+        assert (
+            depth is not None
+        ), "particle's initial depth is None - incompatible with the ParticleData class. Invalid state."
+        assert lon.size == lat.size and lon.size == depth.size, "lon, lat, depth don" "t all have the same lenghts."
```
Comment (erikvansebille) [inline parcels/particledata.py:71]:
```suggestion
        assert lon.size == lat.size and lon.size == depth.size, "lon, lat, depth don't all have the same lenghts."
```

File: parcels/particlefile.py [hunk 8]
```diff
@@ -204,17 +212,19 @@ def write(self, pset, time, indices=None):
             logger.warning("ParticleSet is empty on writing as array at time %g" % time)
             return
 
-        indices_to_write = pset.particledata._to_write_particles(pset.particledata._data, time) if indices is None else indices
+        indices_to_write = (
+            pset.particledata._to_write_particles(pset.particledata._data, time) if indices is None else indices
+        )
```
[the remark this answers is not in the document]
Comment (VeckoTheGecko) [reply parcels/particlefile.py:217]: I agree, but unwieldy ternaries are quite commonplace in the codebase. I'll apply this and save the rest for a future refactoring

File: parcels/field.py [hunk 9]
```diff
@@ -1391,39 +1549,42 @@ def write(self, filename, varname=None):
         varname : str
             Name of the field, to be appended to the filename. (Default value = None)
         """
-        filepath = str(Path(f'{filename}{self.name}.nc'))
+        filepath = str(Path(f"{filename}{self.name}.nc"))
         if varname is None:
             varname = self.name
         # Derive name of 'depth' variable for NEMO convention
-        vname_depth = 'depth%s' % self.name.lower()
+        vname_depth = "depth%s" % self.name.lower()
```
[the remark this answers is not in the document]
Comment (VeckoTheGecko) [reply parcels/field.py:1556]: Yes. Will do the rest of them as part of #1620 pyupgrade

**Comment by VeckoTheGecko** [comment]
@erikvansebille I was wanting to run some tests locally, but got an error about not finding the compiler. I exported `CC` to match `clang --version` but still get the error. Have you seen this before?

```

    def _create_compile_process_(self, cmd, src, log):
        with open(log, "w") as logfile:
            try:
                subprocess.check_call(cmd, stdout=logfile, stderr=logfile)
            except OSError:
>               raise RuntimeError(f"OSError during compilation. Please check if compiler exists: {self._cc}")
E               RuntimeError: OSError during compilation. Please check if compiler exists: arm64-apple-darwin23.6.0

parcels/compilation/codecompiler.py:240: RuntimeError
```

**Comment by erikvansebille** [comment]
> @erikvansebille I was wanting to run some tests locally, but got an error about not finding the compiler. I exported `CC` to match `clang --version` but still get the error. Have you seen this before?
> 
> ```
> 
>     def _create_compile_process_(self, cmd, src, log):
>         with open(log, "w") as logfile:
>             try:
>                 subprocess.check_call(cmd, stdout=logfile, stderr=logfile)
>             except OSError:
> >               raise RuntimeError(f"OSError during compilation. Please check if compiler exists: {self._cc}")
> E               RuntimeError: OSError during compilation. Please check if compiler exists: arm64-apple-darwin23.6.0
> 
> parcels/compilation/codecompiler.py:240: RuntimeError
> ```

No, haven't seen his specific error before. But I normally do `export CC=gcc` when I encounter compiler problems on macOS; that often seems to do the trick

**Pull request merged**